[CI] SharedClusterSnapshotRestoreIT testDataFileFailureDuringRestore failing

The test failed on a PR branch with the following assertion error:

```java
WARNING: Uncaught exception in thread: Thread[elasticsearch[node_s0][generic][T#6],5,TGRP-SharedClusterSnapshotRestoreIT]
java.lang.AssertionError: add restore for [[test-idx][0]] that already has an existing restore
	at __randomizedtesting.SeedInfo.seed([3B52F82051D5C5FD]:0)
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.restoreShard(BlobStoreRepository.java:3003)
	at org.elasticsearch.index.shard.StoreRecovery.lambda$restore$10(StoreRecovery.java:546)
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:136)
	at org.elasticsearch.common.util.concurrent.ListenableFuture.notifyListenerDirectly(ListenableFuture.java:113)
	at org.elasticsearch.common.util.concurrent.ListenableFuture.addListener(ListenableFuture.java:55)
	at org.elasticsearch.common.util.concurrent.ListenableFuture.addListener(ListenableFuture.java:41)
	at org.elasticsearch.action.StepListener.addListener(StepListener.java:108)
	at org.elasticsearch.action.StepListener.whenComplete(StepListener.java:68)
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:545)
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:301)
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:2358)
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:3160)
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:771)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Build scan:**
https://gradle-enterprise.elastic.co/s/qcd6qqtqlbohq/tests/:server:internalClusterTest/org.elasticsearch.snapshots.SharedClusterSnapshotRestoreIT/testDataFileFailureDuringRestore

**Reproduction line:**
`./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.snapshots.SharedClusterSnapshotRestoreIT.testDataFileFailureDuringRestore" -Dtests.seed=3B52F82051D5C5FD -Dtests.locale=nl-BE -Dtests.timezone=Europe/Saratov -Druntime.java=8`

**Applicable branches:**
7.16

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.snapshots.SharedClusterSnapshotRestoreIT&tests.test=testDataFileFailureDuringRestore

**Failure excerpt:**
```
java.lang.Exception: Test abandoned because suite timeout was reached.

  at __randomizedtesting.SeedInfo.seed([3B52F82051D5C5FD]:0)

```

--- a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
+++ b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
@@ -2980,7 +2980,7 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
             final boolean added = ongoingRestores.add(shardId);
             assert added : "add restore for [" + shardId + "] that already has an existing restore";
         }
-        executor.execute(ActionRunnable.wrap(ActionListener.runAfter(restoreListener, () -> {
+        executor.execute(ActionRunnable.wrap(ActionListener.runBefore(restoreListener, () -> {
             final List<ActionListener<Void>> onEmptyListeners;
             synchronized (ongoingRestores) {
                 if (ongoingRestores.remove(shardId) && ongoingRestores.isEmpty() && emptyListeners != null) {

